qiskit/circuit/quantumcircuit.py

```diff
@@ -1205,22 +1205,25 @@ def append(
             cargs = instruction.clbits
         else:
             operation = instruction
+
         # Convert input to instruction
-        if not isinstance(operation, Operation) and not hasattr(operation, "to_instruction"):
-            if issubclass(operation, Operation):
+        if not isinstance(operation, Operation):
+            if hasattr(operation, "to_instruction"):
+                operation = operation.to_instruction()
+                # probably redundant, but check to be sure
+                if not isinstance(operation, Operation):
+                    raise CircuitError("operation.to_instruction() is not an Operation.")
```

Comment:
My only (very minor) suggestion is to remove this comment, as I don't think that we check anywhere else that whenever we call `to_instruction`, the returned values is indeed an `Instruction` (or an `Operation`) and thus can be added onto a quantum circuit.

Reply:
I think in `append` we should probably keep the check, but remove the comment.  I'm 90% sure there's extant code that depends on this behaviour (though it wouldn't have been my choice) - I'm thinking things like people previously `append`ing `Clifford` operations (pre `Operation`) on to circuits.

I'm in favour of removing the comment because it's not really a redundant check - doing the check is consistent with the rest of the logic, and we're already in a "slow" path if we take this route anyway.  `to_instruction` isn't a defined interface (and `instruction` is a pretty generic name), so there's plenty of chance that some object you _might_ expect could be passed to a circuit might expose such a method for other reasons.  For example, there's also `pulse.Instruction` within Qiskit - I don't think there's anything there that defines a `to_instruction` method, but you can certainly see how something similar _might_.

Reply:
Other than that, I already looked at this and am happy to move to merge - I just didn't want to jump in and step on your toes, Sasha.

Reply:
We are on the same page I think. I removed the comment.